### PR TITLE
Return string instead of output pointer in metadata

### PR DESCRIPTION
**Pull request by PragmaTwice**
```diff
-void ExtractNamespaceKey(Slice ns_key, std::string *ns, std::string *key, bool slot_id_encoded);
-void ComposeNamespaceKey(const Slice &ns, const Slice &key, std::string *ns_key, bool slot_id_encoded);
-void ComposeSlotKeyPrefix(const Slice &ns, int slotid, std::string *output);
+[[nodiscard]] std::tuple<Slice, Slice> ExtractNamespaceKey(Slice ns_key, bool slot_id_encoded);
+[[nodiscard]] std::string ComposeNamespaceKey(const Slice &ns, const Slice &key, bool slot_id_encoded);
+[[nodiscard]] std::string ComposeSlotKeyPrefix(const Slice &ns, int slotid);
 
 class InternalKey {
  public:
   ...
-  void Encode(std::string *out);
+  [[nodiscard]] std::string Encode() const;
```

### CODE REVIEW

File: src/storage/redis_metadata.cc [hunk 1]
```diff
@@ -110,41 +109,46 @@ bool InternalKey::operator==(const InternalKey &that) const {
   return version_ == that.version_;
 }
 
-void ExtractNamespaceKey(Slice ns_key, std::string *ns, std::string *key, bool slot_id_encoded) {
+std::tuple<Slice, Slice> ExtractNamespaceKey(Slice ns_key, bool slot_id_encoded) {
   uint8_t namespace_size = 0;
   GetFixed8(&ns_key, &namespace_size);
-  *ns = ns_key.ToString().substr(0, namespace_size);
+  Slice ns(ns_key.data(), namespace_size);
```
Comment (mapleFU) [inline src/storage/redis_metadata.cc]: So this is an optimization that return Slice rather than string?

**Review by mapleFU — DISMISSED** [review]
The code LGTM, lets wait for CI and asan, since it might introduce memory problem...

File: src/storage/redis_metadata.cc (hunk 1 above)
Comment (PragmaTwice) [reply src/storage/redis_metadata.cc]: Yeah, here it is unnecessary to return string. And if it is required, users can construct string manually from slice.

**Review by torwig — APPROVED** [review]
LGTM.

:+1:

**Review by mapleFU — APPROVED** [review]
LGTM, though it increase the risk of lifetine handling, the patch would improve our style(and make some string copy less)

**Pull request merged**